src/map/battleground.cpp

```diff
@@ -222,59 +621,863 @@ int bg_team_get_id(struct block_list *bl)
+					if (queue->teama_members.size() == bg->required_players && queue->teamb_members.size() == bg->required_players) // Enough players have joined
+						bg_queue_on_ready(name, queue);
+
+					return BG_APPLY_ACCEPT;
+				} else {
+					while (!list.empty() && queue->teamb_members.size() < bg->required_players) {
+						struct map_session_data *sd2 = list.back();
+
+						list.pop_back();
+
+						if (!sd2)
+							continue;
+
+						if (!bg_queue_check_joinable(bg, sd2, name))
+							continue;
+
+						sd2->bg_queue = queue;
+						clif_bg_queue_apply_result(BG_APPLY_ACCEPT, name, sd2);
+						clif_bg_queue_apply_notify(name, sd2);
+						queue->teamb_members.insert(queue->teamb_members.begin(), sd2);
+					}
+				}
+
+				return BG_APPLY_ACCEPT;
+			}
+		}
+
+		// Create a new queue if none of the existing ones are joinable for this party/guild
+		queue = nullptr;
+		queue = bg_queue_create(bg);
+		r = rnd() % 2 != 0;
+
+		if (!r) {
+			while (!list.empty() && queue->teama_members.size() < bg->required_players) {
+				struct map_session_data *sd2 = list.back();
+
+				list.pop_back();
+
+				if (!sd2 || sd2->bg_queue)
+					continue;
+
+				if (!bg_queue_check_joinable(bg, sd2, name))
+					continue;
+
+				sd2->bg_queue = queue;
+				clif_bg_queue_apply_result(BG_APPLY_ACCEPT, name, sd2);
+				clif_bg_queue_apply_notify(name, sd2);
+				queue->teama_members.insert(queue->teama_members.begin(), sd2);
+			}
+		} else {
+			while (!list.empty() && queue->teamb_members.size() < bg->required_players) {
+				struct map_session_data *sd2 = list.back();
+
+				list.pop_back();
+
+				if (!sd2 || sd2->bg_queue)
+					continue;
+
+				if (!bg_queue_check_joinable(bg, sd2, name))
+					continue;
+
+				sd2->bg_queue = queue;
+				clif_bg_queue_apply_result(BG_APPLY_ACCEPT, name, sd2);
+				clif_bg_queue_apply_notify(name, sd2);
+				queue->teamb_members.insert(queue->teamb_members.begin(), sd2);
+			}
+		}
+
+		bg_queues.insert(bg_queues.begin(), queue);
+
+		return BG_APPLY_ACCEPT;
+	}
+
+	return BG_APPLY_RECONNECT; // Something went wrong, sends reconnect and then reapply message to client.
+}
+
+/**
+ * Sub function for leaving a Battleground queue
+ * @param sd: Player leaving
+ * @param list: List of players in queue data
+ * @return True on success or false otherwise
+ */
+static bool bg_queue_leave_sub(struct map_session_data *sd, std::vector<map_session_data *> list, std::vector<map_session_data *> listb)
+{
+	for (int i = 0; i < list.size(); i++) {
+		struct map_session_data *player = list.at(i);
+
+		if (player == sd) {
+			if (sd->bg_queue->in_ready_state) {
+				sd->bg_queue->accepted_players = 0;
+				sd->bg_queue->in_ready_state = false;
+				sd->bg_queue_accept_state = false;
+			}
+
+			util::erase_at(list, i);
+
+			if (list.empty() && listb.empty()) { // If there are no players left in the queue, discard it
+				std::vector<std::shared_ptr<s_battleground_queue>>::iterator it = bg_queues.begin();
+
+				while (it != bg_queues.end()) {
```

Comment:
You should replace all of this while loop with an `std::remove_if()`, Like
```suggestion
				std::remove_if(bg_queues.begin(), bg_queues.end(), [](const auto& q) {
		                    if (sd->bg_queue == q) {
				        if (q->tid_requeue && get_timer(q->tid_requeue)) {
					    delete_timer(q->tid_requeue, bg_on_ready_loopback);
					    q->tid_requeue = 0;
			                }
			                it = bg_queues.erase(it);
			            }
				    return true;
				});
```

Reply:
This and the `count_if` I'm 50/50 on. While I understand their uses, might be better to get more opinions on this from other developers.